pkg/manager/internal.go

```diff
@@ -237,40 +258,158 @@ func (cm *controllerManager) AddReadyzCheck(name string, check healthz.Checker)
+				cm.logger.Error(err, "failed to engage cluster with runnable, won't retry", "clusterName", name, "runnable", aware)
+				continue
+			}
+		}
+	}
+}
+
+func (cm *controllerManager) getLogicalCluster(ctx context.Context, name logical.Name) (c *logicalCluster, err error) {
+	// Check if the manager was configured with a logical adapter,
+	// otherwise we cannot retrieve the cluster.
+	if cm.logicalProvider == nil {
+		return nil, fmt.Errorf("manager was not configured with a logical adapter, cannot retrieve %q", name)
+	}
+
+	// Check if the cluster already exists.
+	cm.logicalLock.RLock()
+	c, ok := cm.logicalClusters[name]
+	cm.logicalLock.RUnlock()
+	if ok {
+		return c, nil
+	}
+
+	// Lock the whole function to avoid creating multiple clusters for the same name.
+	cm.logicalLock.Lock()
+	defer cm.logicalLock.Unlock()
+
+	// Check again in case another goroutine already created the cluster.
+	c, ok = cm.logicalClusters[name]
+	if ok {
+		return c, nil
+	}
+
+	// Create a new cluster.
+	var cl cluster.Cluster
+	{
+		// TODO(vincepri): Make this timeout configurable.
+		ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+		defer cancel()
+		var watchErr error
+		if err := wait.PollImmediateUntilWithContext(ctx, 1*time.Second, func(ctx context.Context) (done bool, err error) {
+			cl, watchErr = cm.logicalProvider.Get(ctx, name, cm.defaultClusterOptions, cluster.WithName(name))
+			if watchErr != nil {
+				return false, nil //nolint:nilerr // We want to keep trying.
+			}
+			return true, nil
+		}); err != nil {
+			return nil, fmt.Errorf("failed  create logical cluster %q: %w", name, kerrors.NewAggregate([]error{err, watchErr}))
+		}
+	}
+
+	// We add the Cluster to the manager as a Runnable, which is going to be categorized
+	// as a Cache-backed Runnable.
+	//
+	// Once added, if the manager has already started, it waits for the Cache to sync before returning
+	// otherwise it enqueues the Runnable to be started when the manager starts.
+	if err := cm.Add(cl); err != nil {
+		return nil, fmt.Errorf("cannot add logical cluster %q to manager: %w", name, err)
+	}
+	// Create a new context for the Cluster, so that it can be stopped independently.
+	ctx, cancel := context.WithCancel(context.Background())
+	c = &logicalCluster{
+		Cluster: cl,
+		ctx:     ctx,
+		cancel:  cancel,
+	}
+	cm.logicalClusters[name] = c
+	return c, nil
+}
+
+func (cm *controllerManager) removeLogicalCluster(name logical.Name) error {
+	// Check if the manager was configured with a logical adapter,
+	// otherwise we cannot retrieve the cluster.
+	if cm.logicalProvider == nil {
+		return fmt.Errorf("manager was not configured with a logical adapter, cannot retrieve %q", name)
+	}
+
+	cm.logicalLock.Lock()
+	defer cm.logicalLock.Unlock()
+	c, ok := cm.logicalClusters[name]
+	if !ok {
+		return nil
+	}
+
+	// Disengage all the runnables.
+	for _, aware := range cm.logicalClusterAwareRunnables {
+		if err := aware.Disengage(c.ctx, c); err != nil {
```

Comment:
nit: might make sense to collect error here to try removing it from as many runnables as possible and then return after the loop if we have some errors.